cli/src/commands/run.rs

```diff
@@ -14,15 +14,376 @@
+    }
+    Ok(results)
+}
+
+/// Get the shell to execute in.
+// TODO: use something like `[run].shell`
+fn get_shell_executable_with_first_arg() -> (String, String) {
+    if cfg!(target_os = "windows") {
+        ("cmd".into(), "/c".into())
+    } else {
+        ("/bin/sh".into(), "-c".into())
+    }
+}
+
+/// The result of a single command invocation.
+struct RunJob {
+    /// The old `CommitId` of the commit.
+    old_id: CommitId,
+    /// The new tree generated from the commit.
+    new_tree: Tree,
+    /// Was the tree even modified.
+    dirty: bool,
+}
+
+// TODO: make this more revset/commit stream friendly.
+async fn run_inner<'a>(
+    tx: &WorkspaceCommandTransaction<'a>,
+    sender: Sender<RunJob>,
+    handle: &tokio::runtime::Handle,
+    shell_command: Arc<String>,
+    commits: Arc<Vec<Arc<OnDiskCommit>>>,
+) -> Result<(), RunError> {
+    let mut command_futures = JoinSet::new();
+    for commit in commits.iter() {
+        command_futures.spawn_on(
+            rewrite_commit(
+                // TODO: handle error here
+                tx.base_workspace_helper().base_ignores().unwrap().clone(),
+                ConflictMarkerStyle::Snapshot,
+                commit.clone(),
+                shell_command.clone(),
+            ),
+            handle,
+        );
+    }
+
+    while let Some(res) = command_futures.join_next().await {
+        let done = match res {
+            Ok(rj) => rj?,
+            Err(err) => return Err(RunError::JobFailure(err)),
+        };
+        let should_quit = sender.send(done).is_err();
+        if should_quit {
+            tracing::debug!(
+                ?should_quit,
+                "receiver is no longer available, exiting loop"
+            );
+            break;
+        }
+    }
+    Ok(())
+}
+
+/// Rewrite a single `OnDiskCommit`. The caller is responsible for creating the
+/// final commit.
+async fn rewrite_commit<'a>(
+    base_ignores: Arc<GitIgnoreFile>,
+    conflict_marker_style: ConflictMarkerStyle,
+    stored_commit: Arc<OnDiskCommit>,
+    shell_command: Arc<String>,
+) -> Result<RunJob, RunError> {
+    let (stdout, stderr) =
+        create_output_files(stored_commit.commit.id(), &stored_commit.output_dir)?;
+    // TODO: Later this should take some trait which allows `run` to integrate with
+    // something like Bazels RE protocol.
+    // e.g
+    // ```
+    // let mut executor /* Arc<dyn CommandExecutor> */ = store.get_executor();
+    // let command = executor.spawn(...)?; // RE or separate processes depending on impl.
+    // ...
+    // ```
+    tracing::debug!(
+        "trying to run {shell_command} on commit {id}",
+        id = stored_commit.commit.id().hex(),
+        shell_command = shell_command.as_str()
+    );
+    let (prog, first_arg) = get_shell_executable_with_first_arg();
+    let mut command = tokio::process::Command::new(prog)
+        .arg(first_arg)
+        .arg(shell_command.as_str())
+        // set cwd to the working copy directory.
```

Comment:
(Possibly for later)

For `jj run` command, it is surprising to me that commands are run in the working copy directory instead of the subdirectory I am in. For example, I think it's natural to be in the `cli/src` directory and to run `jj run "mv config.rs somethingelse.rs" -r ...`.

I think this should be governed by an option to `jj run`, and the default should be to run the command in the subdirectory. Until this is implemented, you could make `jj run` vocal about the fact that it's running the commands in the working copy dir.

Reply:
I don't think its for discussion here, as it out of scope for this MVP. Please add it either to #1869 or #3808

Reply:
This TODO is related to this request from @matts1: 
> No mention was made about the working directory, so I'm curious how relative paths would work
IMO, we should set the $JJ_REPO environment variable so that a relative path to the root of the repo can be $JJ_REPO/foo
Maintaining relative paths has value
A unix mount namespace could solve this by literally just overlaying our new virtual file system on top of the real one
But unclear how this'd work on non-unix systems